### PR TITLE
Publish nuget package

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -183,4 +184,8 @@ build_and_release_dotnet_task:
     - ps: SonarScanner.MSBuild.exe end /d:sonar.login="$env:SONAR_TOKEN"
   unit_test_script:
     - echo "Running Unit Tests"
-    - dotnet test ${SOLUTION_DIR} --configuration ${BUILD_CONFIGURATION} --no-build
+    - cd $PROJECT_DIR
+    - dotnet test --configuration ${BUILD_CONFIGURATION} --no-build
+  promote_script:
+    - source cirrus-env QA
```
Comment (costin-zaharia-sonarsource) [inline .cirrus.yml:190]: This is needed only for retrieving the build number.

File: sonar-text-dotnet/nuget.config [hunk 2]
```diff
@@ -1,17 +1,30 @@
...
-    <add key="enabled" value="True" />
-
-    <!-- Automatically check for missing packages during build in Visual Studio -->
-    <add key="automatic" value="True" />
-  </packageRestore>
-
   <packageSources>
     <clear />
-    <add key="slvs_input" value="https://sonarsource.pkgs.visualstudio.com/399fb241-ecc7-4802-8697-dcdd01fbb832/_packaging/slvs_input/nuget/v3/index.json" />
+    <add key="Repox Artifactory (proxy)" value="https://repox.jfrog.io/artifactory/api/nuget/nuget" />
```
Comment (costin-zaharia-sonarsource) [inline sonar-text-dotnet/nuget.config:5]: We will no longer use the azure feed. The plan is to move away from it.

I've changed this to point to repox and do the package signature validation.

File: sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1 [hunk 3]
```diff
@@ -6,19 +6,24 @@ function CheckIfSuccessful($StepName) {
...
+jf rt ping # Check if the artifactory configuration is successfull
+
+Write-Host "Setting up artifactory dotnet configuration"
+jf dotnet-config --global --server-id-resolve repox --repo-resolve $nugetFeed
+
+$signAssembly = "$env:CIRRUS_BRANCH" -eq "master" -or "$env:CIRRUS_BRANCH".startsWith("branch-")
+Write-Host "Should the assembly be signed: ${signAssembly}"
+
+Write-Host "Building project"
+jf dotnet build `
```
Comment (costin-zaharia-sonarsource) [inline sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1:26]: For the rest of `dotnet` commands, excepting signing, we will use the `jf`` wrapper. This collects build information that will later be published to artifactory.

File: sonar-text-dotnet/pipeline_scripts/promote.ps1 [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+Write-Host "Setting up artifactory authentication for promotion"
+jf config add "repox" --url https://repox.jfrog.io --access-token $env:ARTIFACTORY_PROMOTE_TOKEN --overwrite
+jf rt ping # Check if the artifactory configuration is successfull
+
+$isBranchBuild = "$env:CIRRUS_BRANCH" -eq "master" -or "$env:CIRRUS_BRANCH".startsWith("branch-")
+$sourceRepo = "sonarsource-nuget-qa"
+$destinationRepo = if ($isBranchBuild) { "sonarsource-nuget-public" } else { "sonarsource-nuget-dev" }
+
+Write-Host "Promoting the NuGet package from ${sourceRepo} to ${destinationRepo}"
+jf rt build-promote $env:ARTIFACTORY_BUILD_NAME $env:BUILD_NUMBER $destinationRepo --source-repo $sourceRepo
```
Comment (costin-zaharia-sonarsource) [inline sonar-text-dotnet/pipeline_scripts/promote.ps1:12]: The promoted versions can be found here:
https://repox.jfrog.io/ui/packages/nuget:%2F%2FSonarLint.Secrets.DotNet?name=SonarLint.Secrets.DotNet&type=packages

File: sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1 [hunk 5]
```diff
@@ -6,19 +6,24 @@ function CheckIfSuccessful($StepName) {
...
+$nugetFeed = "https://repox.jfrog.io/artifactory/api/nuget/nuget"
+
+Write-Host "Setting up artifactory authentication for QA deploy"
+jf config add "repox" --url https://repox.jfrog.io --access-token $env:ARTIFACTORY_DEPLOY_PASSWORD
+jf rt ping # Check if the artifactory configuration is successfull
+
+Write-Host "Setting up artifactory dotnet configuration"
+jf dotnet-config --global --server-id-resolve repox --repo-resolve $nugetFeed
+
+$signAssembly = "$env:CIRRUS_BRANCH" -eq "master" -or "$env:CIRRUS_BRANCH".startsWith("branch-")
```
Comment (csaba-sagi-sonarsource) [inline sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1:22]: For consistency, I suggest adding the curly brackets as all other places use them.
```suggestion
$signAssembly = "${env:CIRRUS_BRANCH}" -eq "master" -or "${env:CIRRUS_BRANCH}".startsWith("branch-")
```

File: sonar-text-dotnet/pipeline_scripts/promote.ps1 [hunk 6]
```diff
@@ -0,0 +1,16 @@
+Set-Location "${env:CIRRUS_WORKING_DIR}\\sonar-text-dotnet"
+
+Write-Host "Setting up artifactory authentication for promotion"
+jf config add "repox" --url https://repox.jfrog.io --access-token $env:ARTIFACTORY_PROMOTE_TOKEN --overwrite
+jf rt ping # Check if the artifactory configuration is successfull
+
+$isBranchBuild = "$env:CIRRUS_BRANCH" -eq "master" -or "$env:CIRRUS_BRANCH".startsWith("branch-")
```
Comment (csaba-sagi-sonarsource) [inline sonar-text-dotnet/pipeline_scripts/promote.ps1:7]:
```suggestion
$isMasterOrBranchBuild = "$env:CIRRUS_BRANCH" -eq "master" -or "$env:CIRRUS_BRANCH".startsWith("branch-")
```

**Review by csaba-sagi-sonarsource — APPROVED** [review]
I have two nitpick comments, otherwise good to be merged.

File: sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1 (hunk 5 above)
Comment (costin-zaharia-sonarsource) [reply sonar-text-dotnet/pipeline_scripts/build_and_pack_solution.ps1]: Ah, they are not needed. I think we should remove them from everywhere they aren't needed.

**Pull request merged**